### DISCUSSION TITLE
can not find modules from pygubu

### DISCUSSION THREAD

**Discussion by NewFlow** [pyinstaller/pyinstaller#6104] [Help] [answered]
Hello. I encountered a problem after I used Pyinstaller to build exe files. I had pygubu module in the .py file, and I had error message of: "No module named 'pygubu.builder.widgets'"; "No module named 'tkscrolledframe'" when I tried to run the .exe file, although I don't have any problem to run the .py file. I did some research, and I tried to use --hidden-import=gypubu.builder.widgets and --hidden-import=tkinter to build, it didn't work. I also tried modifying the .spec file, and added
  hidden_imports = [
    'pygubu.builder.tkstdwidgets',
    'pygubu.builder.ttkstdwidgets',
    'pygubu.builder.widgets.dialog',
    'pygubu.builder.widgets.editabletreeview',
    'pygubu.builder.widgets.scrollbarhelper',
    'pygubu.builder.widgets.scrolledframe',
    'pygubu.builder.widgets.tkscrollbarhelper',
    'pygubu.builder.widgets.tkscrolledframe',
    'pygubu.builder.widgets.pathchooserinput',
]
in the .spec file. 
I still got the problem. What else did I miss? 
Below is the code:

import os
import tkinter as tk
import tkinter.ttk as ttk
import pygubu
from pygubu.builder import tkstdwidgets

PROJECT_PATH = os.path.abspath(os.path.dirname(__file__))
PROJECT_UI = os.path.join(PROJECT_PATH, "test.ui")

class TestApp:
    def __init__(self, master=None):
        self.builder = builder = pygubu.Builder()
        builder.add_resource_path(PROJECT_PATH)
        builder.add_from_file(PROJECT_UI)
        self.mainwindow = builder.get_object('mainwindow', master)
        builder.connect_callbacks(self)
    

    def run(self):
        self.mainwindow.mainloop()

if __name__ == '__main__':
    root = tk.Tk()
    app = TestApp(root)
    app.run()

In the test.ui, there is only a button and a scrolledframe. 

Thanks!

**Comment by rokm** [comment, marked as answer]
Can you try adding `--collect-submodules pygubu` to your PyInstaller command? If that doesn't help, please paste whole error traceback.

**Reply by NewFlow** [reply]
Magically it works! Thank you very much. Can you tell me what went wrong with the 2 methods I used in the original post?

**Reply by rokm** [reply]
When you added those `hidden_imports` to the .spec file, did you pass the hidden_imports variable to the `Analysis()`?

**Reply by NewFlow** [reply]
No. I didn't. I didn't see that when I read the solutions on similar problems from other people. Do you mind to show me where I can find the document for this? Thanks!

**Reply by rokm** [reply]
You were probably reading the solutions where custom hooks were created; in those, it is enough to just create `hiddenimports` variable and list the modules. If you edit the .spec file, you need to pass any additional `hiddenimports`, `binaries` and/or `datas` via the equivalent arguments of the `Analysis()` (if you check your .spec after using `--collect-submodules`, you will see that it does exactly that).

The spec files should be documented [here](https://pyinstaller.readthedocs.io/en/latest/spec-files.html), although it seems that only `datas` and `binaries` are actually documented...

**Reply by NewFlow** [reply]
I see. At the top of .spec file, the hidden_imports is a variable. The system wouldn't read it directly. I have to put the variable in the Analysis() to make it work. However, I remember I tried to modify the hiddenimports=[] in the Analysis() directly before, and it didn't work either. Did I miss anything else for that too?

**Reply by rokm** [reply]
Perhaps you missed some modules that are now included in the list returned by `collect_submodules`? It's hard to say without knowing what exactly you did, and what the error message was...
